src/ch/ch_process.c

```diff
@@ -372,6 +374,95 @@ virCHProcessSetupVcpus(virDomainObjPtr vm)
     return 0;
 }
 
+/**
+ * chProcessNetworkPrepareDevices
+ */
+static int
+chProcessNetworkPrepareDevices(virCHDriverPtr driver, virDomainObjPtr vm)
+{
+    virDomainDefPtr def = vm->def;
+    size_t i;
+    virCHDomainObjPrivatePtr priv = vm->privateData;
+    g_autoptr(virConnect) conn = NULL;
+    char **tapfdName = NULL;
+    int *tapfd = NULL;
+    size_t tapfdSize = 0;
+
+    for (i = 0; i < def->nnets; i++) {
+        virDomainNetDefPtr net = def->nets[i];
+        virDomainNetType actualType;
+
+        /* If appropriate, grab a physical device from the configured
+         * network's pool of devices, or resolve bridge device name
+         * to the one defined in the network definition.
+         */
+        if (net->type == VIR_DOMAIN_NET_TYPE_NETWORK) {
+            if (!conn && !(conn = virGetConnectNetwork())){
+                return -1;}
+            if (virDomainNetAllocateActualDevice(conn, def, net) < 0){
+                return -1;}
+        }
+
+        actualType = virDomainNetGetActualType(net);
+        if (actualType == VIR_DOMAIN_NET_TYPE_HOSTDEV &&
+            net->type == VIR_DOMAIN_NET_TYPE_NETWORK) {
+            /* Each type='hostdev' network device must also have a
+             * corresponding entry in the hostdevs array. For netdevs
+             * that are hardcoded as type='hostdev', this is already
+             * done by the parser, but for those allocated from a
+             * network / determined at runtime, we need to do it
+             * separately.
+             */
+            virDomainHostdevDefPtr hostdev = virDomainNetGetActualHostdev(net);
+            virDomainHostdevSubsysPCIPtr pcisrc = &hostdev->source.subsys.u.pci;
+
+            if (virDomainHostdevFind(def, hostdev, NULL) >= 0) {
+                virReportError(VIR_ERR_INTERNAL_ERROR,
+                               _("PCI device %04x:%02x:%02x.%x "
+                                 "allocated from network %s is already "
+                                 "in use by domain %s"),
+                               pcisrc->addr.domain, pcisrc->addr.bus,
+                               pcisrc->addr.slot, pcisrc->addr.function,
+                               net->data.network.name, def->name);
+                return -1;
+            }
+            if (virDomainHostdevInsert(def, hostdev) < 0)
+                return -1;
+        } else if (actualType == VIR_DOMAIN_NET_TYPE_USER ) {
+            virReportError(VIR_ERR_INTERNAL_ERROR,
+              _("VIR_DOMAIN_NET_TYPE_USER is not a supported Network type"));
+         } else if (actualType == VIR_DOMAIN_NET_TYPE_NETWORK ) {
+            tapfdSize = net->driver.virtio.queues;
+            if (!tapfdSize)
+                tapfdSize = 2; //This needs to be at least 2, based on
+                // https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/docs/networking.md
+
+            if (VIR_ALLOC_N(tapfd, tapfdSize) < 0 ||
+                VIR_ALLOC_N(tapfdName, tapfdSize) < 0)
```

Comment:
Should use the glib allocation call, [g_new](https://developer.gnome.org/glib/stable/glib-Memory-Allocation.html#g-new) in this case I think.

Reply:
Given you are memsetting below, maybe g_new0 instead.

Reply:
Avoids the need to test for allocation failure as well.

Reply:
Qemu is using the same pattern. do we need to change it?

Reply:
Yea, qemu is grandfathered in. They are switching things out over time but no new code is supposed to use the VIR_ALLOC/FREE parts from my understanding.